Makefile

```diff
@@ -0,0 +1,276 @@
+# not found on the system path.
+_program_%: FORCE
+	@_=$(or $(shell which $* 2> /dev/null),$(error `$*` command not found. Please install `$*` and try again))
+
+# Helper target for declaring required environment variables.
+#
+# For example,
+#   `my_target`: | _var_PARAMETER`
+#
+# will fail before running `my_target` if the variable `PARAMETER` is not declared.
+_var_%: FORCE
+	@_=$(or $($*),$(error `$*` is a required parameter))
+
+_tag: | _var_VERSION
+	make --no-print-directory -B README.md
+	git commit -am "Tagging release $(VERSION)"
+	git tag -a $(VERSION) $(if $(NOTES),-m '$(NOTES)',-m $(VERSION))
+.PHONY: _tag
+
+_push: | _var_VERSION
+	git push origin $(VERSION)
+	git push origin master
+.PHONY: _push
+
+## ------ Commmands -----------
+
+TARGET_MAX_CHAR_NUM=20
+## Show help
+help:
+	@echo ''
+	@echo 'Usage:'
+	@echo '  ${YELLOW}make${RESET} ${GREEN}<target>${RESET}'
+	@echo ''
+	@echo 'Targets:'
+	@awk '/^[a-zA-Z\-\_0-9]+:/ { \
+		helpMessage = match(lastLine, /^## (.*)/); \
+		if (helpMessage) { \
+			helpCommand = substr($$1, 0, index($$1, ":")-1); \
+			helpMessage = substr(lastLine, RSTART + 3, RLENGTH); \
+			printf "  ${YELLOW}%-$(TARGET_MAX_CHAR_NUM)s${RESET} ${GREEN}%s${RESET}\n", helpCommand, helpMessage; \
+		} \
+	} \
+	{ lastLine = $$0 }' \
+	$(MAKEFILE_LIST)
+
+## Install dependencies.
+setup: \
+	pre_setup \
+	install_rust \
+	install_rustup \
+	install_rust_toolchain \
+	build_rust_dependencies
+
+# check_for_homebrew \
+# update_homebrew \
+
+pull_request: \
+	test \
+	codecov_upload \
+	danger
+
+pre_setup:
+	$(info Project setup…)
+
+check_for_rust:
+	$(info Checking for Rust…)
+
+ifeq ($(RUST),)
+	$(error Rust is not installed.)
+endif
+
+check_for_rustup:
+	$(info Checking for Rustup…)
+
+ifeq ($(RUSTUP),)
+	$(error Rust is not installed.)
+endif
+
+check_for_ruby:
+	$(info Checking for Ruby…)
+
+ifeq ($(RUBY),)
+	$(error Ruby is not installed.)
+endif
+
+check_for_homebrew:
+	$(info Checking for Homebrew…)
+
+ifeq ($(HOMEBREW),)
+	$(error Homebrew is not installed)
+endif
+
+update_homebrew:
+	$(info Updating Homebrew…)
+
+	brew update
+
+install_swift_lint:
+	$(info Install swiftlint…)
+
+	brew unlink swiftlint || true
+	brew install swiftlint
+	brew link --overwrite swiftlint
+
+install_bundler_gem:
+	$(info Checking and installing bundler…)
+
+ifeq ($(BUNDLER),)
+	gem install bundler -v '~> 1.17'
+else
+	gem update bundler '~> 1.17'
+endif
+
+install_ruby_gems:
+	$(info Installing Ruby gems…)
+
+	bundle install
+
+pull:
+	$(info Pulling new commits…)
+
+	git stash push || true
+	git pull
+	git stash pop || true
+
+## -- Source Code Tasks --
+
+## Pull upstream and update 3rd party frameworks
+# update: pull submodules build_rust_dependencies
+update: submodules build_rust_dependencies
+
+submodules:
+	$(info Updating submodules…)
+
+	git submodule update --init --recursive
+
+build_rust_dependencies:
+	$(info Building Rust dependencies…)
+
+	curl https://sh.rustup.rs -sSf | sh
```

Comment:
Why are we reinstalling rustc here? There’s already code for this below.

Reply:
rustup, the package manager, is a separate install in my testing

Reply:
this makefile is a WIP too, changes welcome or we can edit it as we go

Reply:
But you’re running that rustup install request twice

Reply:
oh, this was a mistake. i was testing this part. i pushed an older version pulling branches apart. I'll fix

Reply:
This is now fixed.

Really i should test if each step is required, I have a more complex shell script that has functions to do that easily, but for now, this works for devs.